transport/src/main/java/io/netty/bootstrap/ChannelInitializerExtensions.java

```diff
@@ -0,0 +1,115 @@
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.bootstrap;
+
+import io.netty.util.internal.SystemPropertyUtil;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.ServiceLoader;
+
+/**
+ * The configurable facade that decides what {@link ChannelInitializerExtension}s to load and where to find them.
+ */
+abstract class ChannelInitializerExtensions {
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(ChannelInitializerExtensions.class);
+    private static volatile ChannelInitializerExtensions implementation;
+
+    private ChannelInitializerExtensions() {
+    }
+
+    /**
+     * Get the implementation where we have no extensions.
+     */
+    static ChannelInitializerExtensions empty() {
+        return EmptyExtensions.EMPTY;
+    }
+
+    /**
+     * Get the configuration extensions, which may be the same as {@link #empty()} if the
+     * {@code io.netty.bootstrap.extensions} system property is {@code empty} or {@code false}.
+     * <p>
+     * By default, we pick an implementation that finds extensions through
+     * {@linkplain ServiceLoader#load(Class) service loading}.
+     */
+    static ChannelInitializerExtensions getExtensions() {
+        ChannelInitializerExtensions impl = implementation;
+        if (impl == null) {
+            synchronized (ChannelInitializerExtensions.class) {
+                impl = implementation;
+                if (impl != null) {
+                    return impl;
+                }
+                String extensionProp = SystemPropertyUtil.get("io.netty.bootstrap.extensions");
+                logger.debug("-Dio.netty.bootstrap.extensions: {}", extensionProp);
+                if ("serviceload".equalsIgnoreCase(extensionProp)) {
+                    impl = new ServiceLoadingExtensions();
+                } else {
+                    impl = empty();
+                }
+                implementation = impl;
+            }
+        }
+        return impl;
+    }
+
+    /**
+     * Get the list of available extensions. The list is unmodifiable.
+     */
+    abstract List<ChannelInitializerExtension> extensions();
+
+    private static final class EmptyExtensions extends ChannelInitializerExtensions {
```

Comment:
nit: EmptyExtensions class may be replaced with just static field: static final ChannelInitializerExtensions EMPTY_EXTENSIONS = new ChannelInitializerExtensions() ...

Reply:
A class initializer should not create sub-class instances, as that can lead to class loader deadlocks.